### PR TITLE
ling3 support + KIMI  PATCH 

### PR DESCRIPTION
**Pull request by ved1beta**
# Description
Ling 3.0 — there's an in-tree modeling copy at all (the published remote code hardcodes eager attention, so flash_attention_2 would silently train without a causal mask), then the packing and the registration checklist.

Kimi — separates the two changes : the fla 0.4.1 fused_kda_gate signature fix,
 and the extraction of the remote-code redirect into the shared redirect_dynamic_modules helper that Ling reuses. 

 second one is the reason patch_kimi_linear.py went 90 → 15 lines,

## Motivation and Context
https://huggingface.co/inclusionAI/Ling-3.0-flash

## How has this been tested?
configs 
(currently running kimi configs for verification) 
## AI Usage Disclaimer
claude help with runs 

## Summary by CodeRabbit

* **New Features**
  * Added support for fine-tuning Ling 3.0 models, including Flash and Tiny variants.
  * Added Ling 3.0 guides, training examples, quantization, LoRA, and dataset configuration guidance.
  * Added support for packed-sequence training and multipack workflows.

* **Bug Fixes**
  * Improved Kimi Delta Attention gating and remote-code handling.

* **Tests**
  * Added coverage for Ling 3.0 modeling, attention, masking, checkpoint conversion, and document isolation.

### CODE REVIEW

File: examples/ling/ling-3.0-flash-qlora.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# See examples/ling/README.md for architecture notes and requirements.
```
Comment (NanoCode012) [inline examples/ling/ling-3.0-flash-qlora.yaml]: Redundant

File: examples/ling/ling-3.0-flash-qlora.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+trust_remote_code: true
+
+chat_template: tokenizer_default
+datasets:
+  - path: fozziethebeat/alpaca_messages_2k_test
+    type: chat_template
+    split: train
+
+dataset_prepared_path: last_run_prepared
```
Comment (NanoCode012) [inline examples/ling/ling-3.0-flash-qlora.yaml]: Let’s default to not set this

File: examples/ling3/ling-3.0-flash-qlora.yaml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+sample_packing: true
+pad_to_sequence_len: true
+
+load_in_4bit: true
+quantize_moe_experts: true
+adapter: qlora
+lora_r: 16
+lora_alpha: 32
+lora_dropout: 0.0
+# q_lora_rank is null on flash, so its multi-latent layers reuse the plain q_proj name
```
Comment (NanoCode012) [inline examples/ling3/ling-3.0-flash-qlora.yaml]: Could you elaborate what the effects mean?

File: examples/ling/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+family by inclusionAI. Most layers use Kimi Delta Attention (KDA) and every `layer_group_size`-th
+one uses multi-latent attention (MLA), so the KV cache stays small at long context; the FFNs are
+fine-grained MoE with a shared expert and sigmoid `noaux_tc` routing.
+
+| Model | Total / active | Experts | Layers | KDA : MLA |
+|-------|----------------|---------|--------|-----------|
+| [Ling-3.0-tiny](https://huggingface.co/inclusionAI/Ling-3.0-tiny) | 7.9B / 1.3B | 128 | 24 | 18 : 6 |
+| [Ling-3.0-flash](https://huggingface.co/inclusionAI/Ling-3.0-flash) | 124B / 5.1B | 512 | 42 (+1 MTP) | 35 : 7 |
+
+**Note:** Axolotl trains Ling 3.0 with its own modeling code. The published remote code is
```
Comment (NanoCode012) [inline examples/ling/README.md]: I think mentioning that we use custom modeling code is fine, we can keep details sparse at the top, and elaborate later below

File: examples/ling3/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    ```
+
+   Ling-3.0-flash needs multiple GPUs; its config quantizes both the linear layers and the expert
+   tensors:
+
+    ```bash
+    axolotl train examples/ling/ling-3.0-flash-qlora.yaml
+    ```
+
+VRAM for these configs has not been measured yet — please report what you see.
```
Comment (NanoCode012) [inline examples/ling3/README.md]: Todo

File: src/axolotl/model_support/kimi_linear/modeling_kimi.py [hunk 6]
```diff
@@ -564,7 +564,8 @@ def forward(
             cu_seqlens=cu_seqlens,
         )
         g = self.f_b_proj(self.f_a_proj(hidden_states))
-        g = fused_kda_gate(g, self.A_log, self.head_dim, g_bias=self.dt_bias)
+        g = rearrange(g, "... (h d) -> ... h d", d=self.head_dim)
+        g = fused_kda_gate(g, self.A_log, self.dt_bias)
```
Comment (NanoCode012) [inline src/axolotl/model_support/kimi_linear/modeling_kimi.py:568]: I guess this is because the api changes? Can you make sure we pull the correct updated dependency on any related doc?

File: src/axolotl/model_support/kimi_linear/patch_kimi_linear.py [hunk 7]
```diff
@@ -1,95 +1,20 @@
...
-    Apply all Kimi patches.
-    Must be called BEFORE loading config/tokenizer/model.
-    """
-    _patch_get_class_in_module()
+    """Apply all Kimi patches. Must be called BEFORE loading config/tokenizer/model."""
+    redirect_dynamic_modules(KIMI_PATCH_PACKAGE, KIMI_MODULES)
     LOG.info("Kimi patches applied successfully!")
 
 
-# Keep these for backward compatibility if needed
```
Comment (NanoCode012) [inline src/axolotl/model_support/kimi_linear/patch_kimi_linear.py:92]: Should we keep this?

File: README.md [hunk 8]
```diff
@@ -31,7 +31,7 @@
 ## 🎉 Latest Updates
 
 - 2026/08:
-  - New model support has been added in Axolotl for [Muse Glimmer](https://docs.axolotl.ai/docs/models/muse-glimmer.html), [North Micro Vision Instruct](https://docs.axolotl.ai/docs/models/cohere-north-micro-vision-instruct.html) and [Shieldstral](https://docs.axolotl.ai/docs/models/shieldstral.html).
+  - New model support has been added in Axolotl for [Ling 3.0](https://docs.axolotl.ai/docs/models/ling.html), [Muse Glimmer](https://docs.axolotl.ai/docs/models/muse-glimmer.html), [North Micro Vision Instruct](https://docs.axolotl.ai/docs/models/cohere-north-micro-vision-instruct.html) and [Shieldstral](https://docs.axolotl.ai/docs/models/shieldstral.html).
```
Comment (NanoCode012) [inline README.md]: Maybe link needs to be ling3 following their naming?

**Review by NanoCode012 — COMMENTED** [review]
First review pass

File: src/axolotl/model_support/kimi_linear/modeling_kimi.py (hunk 6 above)
Comment (ved1beta) [reply src/axolotl/model_support/kimi_linear/modeling_kimi.py:568]: yes ` fla 0.4.1 ` is fused_kda_gate(g, A_log, dt_bias=None, output_dtype=torch.float32) and expects g pre-shaped [..., H, K];

**Review by NanoCode012 — APPROVED** [review]
